transformer_engine/pytorch/transformer.py

```diff
@@ -211,14 +191,267 @@ def forward(
+        attn_mask_type: str = "causal",
+    ) -> None:
+        super().__init__()
+
+        if "dev" not in _flash_attn_version:
+            raise ImportError(
+                'Please install correct version of flash-attn with ' \
+                'pip install git+https://github.com/ksivaman/flash-attention.git@hopper. ' \
+                'If running on Hopper, ' \
+                'please install from source with compute capability 9.0.')
+        assert (
+            attn_mask_type == "causal"
+            ), 'FlashAttention currently only supports causal attention mask.'
+        assert (
+            attention_softmax_in_fp32
+            ), 'FlashAttention currently only supports softmax compute in fp32.'
+
+        self.attn_causal_mask = attn_mask_type == "causal"
+        self.norm_factor = norm_factor
+        self.attention_dropout_ctx = attention_dropout_ctx
+        self.attention_dropout = attention_dropout
+        self.layer_number = layer_number
+        self.apply_query_key_layer_scaling = apply_query_key_layer_scaling
+
+    def forward(
+        self,
+        query_layer: torch.Tensor,
+        key_layer: torch.Tensor,
+        value_layer: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """flash-attn fprop"""
+
+        assert (
+            (query_layer.dtype in [torch.float16, torch.bfloat16])
+            and (key_layer.dtype in [torch.float16, torch.bfloat16])
+            and (value_layer.dtype in [torch.float16, torch.bfloat16])
+            ), 'FlashAttention currently only supports FP16 and BF16.'
+        assert (
+            query_layer.is_cuda and key_layer.is_cuda and value_layer.is_cuda
+            ), 'FlashAttention currently only supports CUDA tensors.'
+        assert (
+            attention_mask is None
+        ), 'FlashAttention currently does not support attention mask.'
+
+        query_layer, key_layer, value_layer = [x.transpose(0,1).contiguous()
+                       for x in (query_layer, key_layer, value_layer)]
+
+        batch_size, seqlen = query_layer.shape[0], query_layer.shape[1]
+
+        # [b, sq, np, hn]
+        query_layer, key_layer, value_layer = [
+            x.view(x.shape[0] * x.shape[1], *x.shape[2:])
+            for x in [query_layer, key_layer, value_layer]
+        ]
+
+        max_seqlen = seqlen
+        cu_seqlens = torch.arange(
+            0,
+            (batch_size + 1) * seqlen,
+            step=seqlen,
+            dtype=torch.int32,
+            device=query_layer.device)
+
+        with self.attention_dropout_ctx():
+            output = flash_attn_unpadded_func(
+                query_layer, key_layer, value_layer, cu_seqlens, cu_seqlens, max_seqlen, max_seqlen,
+                self.attention_dropout if self.training else 0.0,
+                softmax_scale=1.0/self.norm_factor, causal=self.attn_causal_mask
+            )
+
+        # [b, sq, np, hn]
```

Comment:
Wrong comment or wrong code?

Reply:
The dimension comments are to indicate shape before not after

Reply:
Well, then the code is wrong, since to get that you would need to do transpose, not view, right?

Reply:
You are actually right, everything is wrong here; comment (input shape) and the view. I couldn't catch in my numerical tests since I'm running with mbs=1, sigh